autoparallel/graph_pp_runner.py

```diff
@@ -0,0 +1,285 @@
+import itertools
+import re
+from dataclasses import dataclass
+from typing import Any, Callable, cast, Optional, Union
+
+import torch
+import torch.fx as fx
+from torch.distributed.pipelining.schedules import (
+    _Action,
+    _PipelineContext,
+    _PipelineScheduleRuntime,
+    _wait_batch_p2p,
+)
+from torch.distributed.pipelining.stage import (
+    _normalize_model_output_as_tuple,
+    _PipelineStage,
+)
+from torch.distributed.tensor import DTensor
+
+
+@dataclass
+class GraphCallables:
+    forward: fx.GraphModule
+    backward: fx.GraphModule
+    backward_inputs: Optional[fx.GraphModule] = None
+    backward_weights: Optional[fx.GraphModule] = None
+    unshard: Optional[fx.GraphModule] = None
+    reduce_grad: Optional[fx.GraphModule] = None
+
+
+@dataclass
+class GraphMeta:
+    num_mutate_inputs: int
+    num_user_outputs: int
+    num_symints_saved_for_bw: int
+
+
+class GraphPipelineStage(_PipelineStage):
+    def __init__(
+        self,
+        submodule: torch.nn.Module,
+        graph_callables: GraphCallables,
+        graph_meta: GraphMeta,
+        stage_index: int,
+        num_stages: int,
+        device: torch.device,
+        input_args: Optional[Union[torch.Tensor, tuple[torch.Tensor, ...]]] = None,
+        output_args: Optional[Union[torch.Tensor, tuple[torch.Tensor, ...]]] = None,
+        group: Optional[torch.distributed.ProcessGroup] = None,
+        dw_builder: Optional[Callable[[], Callable[..., None]]] = None,
+    ):
+        super().__init__(
+            submodule=submodule,
+            stage_index=stage_index,
+            num_stages=num_stages,
+            device=device,
+            input_args=input_args,
+            output_args=output_args,
+            group=group,
+            dw_builder=dw_builder,
+        )
+        self.graph_callables = graph_callables
+        self.graph_meta = graph_meta
+
+
+def _run_forward_microbatch(stage: GraphPipelineStage, *args) -> tuple[Any, Any]:
+    stage_mod = stage.submod
+    stage_graphs = stage.graph_callables
+    stage_graph_meta = stage.graph_meta
+    params_and_buffers = [
+        v.to_local() if isinstance(v, DTensor) else v
+        for k, v in itertools.chain(
+            dict(stage_mod.named_parameters(remove_duplicate=False)).items(),
+            dict(stage_mod.named_buffers(remove_duplicate=False)).items(),
+        )
+    ]
+    boxed_args = [
+        *params_and_buffers,
+        *args,
+    ]
+    del params_and_buffers
+    fw_module = stage_graphs.forward
+    fw_outputs = torch.fx.Interpreter(fw_module).boxed_run(boxed_args)
```

Comment:
All of these comments are for after this merges

For efficiency, don't use interpreter here